TL/CUDA: Linear Broadcast for GPU

ikryukov opened the pull request:
## What
Linear CUDA Broadcast implementation with Active set feature support.

## Why ?
- Functional improvement, parity with others communication libraries.
- Ability to place many ranks on single GPU
- No GPU blocking, communication initiated from host
- Active set might be used to emulate P2P send/receive on top of broadcast collective

## How ?
Naive approach where root rank writes data to own shared buffer and others ranks read from it through NVLink (bi-copy).
The algorithm enables the initiation of multiple simultaneous peer-to-peer (P2P) communications, constrained only by user-defined tags. During its execution, the algorithm identifies the first available free barrier and acquires it. This barrier, along with associated resources, is then utilized as a scratch buffer to facilitate efficient operations.

ikryukov commented:
Configuration string:
` ./configure --with-ucx=$HPCX_UCX_DIR --with-cuda=/usr/local/cuda --with-mpi=$HPCX_MPI_DIR --enable-gtest --prefix=$PWD/install --with-nvcc-gencode="-gencode=arch=compute_80,code=sm_80" --enable-debug`
Run string: 
`mpirun  --mca coll ^hcoll --mca coll_ucc_enable 0 -x LD_LIBRARY_PATH=/home/ikryukov/work/ucc/install/lib:$LD_LIBRARY_PATH -x UCC_CLS=basic -x UCC_TLS=ucp,cuda -x xUCC_LOG_LEVEL=info -x UCC_TL_CUDA_LOG_LEVEL=debug -x UCC_LOG_LEVEL=info -x UCC_CONFIG_FILE= -np 2 ./install/bin/ucc_test_mpi -c bcast --teams world  -M cuda -O 0`

swx-jenkins3 commented:
Can one of the admins verify this patch?

samnordmann commented on src/components/tl/cuda/bcast/bcast_linear.c:
```diff
@@ -0,0 +1,303 @@
+    ucc_tl_cuda_task_t *task;
+    ucc_status_t        status;
+
+    if (ucc_unlikely(!ucc_tl_cuda_team_topo_is_fully_conntected(team->topo) ||
```
I understand it is not related to the pr though... but I'm pointing to the typo anyway
```suggestion
    if (ucc_unlikely(!ucc_tl_cuda_team_topo_is_fully_connected(team->topo) ||
```

ikryukov replied:
fixed typo

samnordmann commented on src/components/tl/cuda/bcast/bcast_linear.c:
```diff
@@ -0,0 +1,303 @@
+{
+    ucc_tl_cuda_task_t *task = ucc_derived_of(coll_task, ucc_tl_cuda_task_t);
+    ucc_tl_cuda_team_t *team = TASK_TEAM(task);
+    ucc_coll_args_t *   args = &TASK_ARGS(task);
```
alignment of "*", here and in other places

ikryukov replied:
Agree, but I use provided clang-format and it did this style, maybe it depends on clang-format version?

samnordmann commented on src/components/tl/cuda/bcast/bcast_linear.c:
```diff
@@ -0,0 +1,303 @@
+                         : task->bcast_linear.size -
+                  (task->bcast_linear.step - 1) * half_scratch_size;
+    size_t offset_buff = task->bcast_linear.step * half_scratch_size;
+
```
extra blank line

samnordmann commented on src/components/tl/cuda/bcast/bcast_linear.c:
```diff
@@ -0,0 +1,303 @@
+    switch (task->bcast_linear.stage) {
+    case STAGE_SYNC:
+        if (ucc_tl_cuda_get_sync(task) != UCC_OK) {
+            task->super.status = UCC_INPROGRESS;
```
redundant with L96

samnordmann commented on src/components/tl/cuda/bcast/bcast_linear.c:
```diff
@@ -0,0 +1,303 @@
+    set_rank_step(task, trank, 0, 0);
+    ucc_memory_cpu_store_fence();
+    status = ucc_tl_cuda_shm_barrier_start(UCC_TL_TEAM_RANK(team), task->bar);
+    if (ucc_unlikely(status != UCC_OK)) {
+        goto exit_err;
+    }
+
+    return UCC_OK;
+
+exit_err:
```
```suggestion
```
I understand it comes from the codebase though

ikryukov replied:
removed lines and check format again

samnordmann replied:
what about that?

ikryukov replied:
done

samnordmann commented on src/components/tl/cuda/bcast/bcast_linear.c:
```diff
@@ -0,0 +1,303 @@
+                    return;
+                }
+            } else {
+                ucc_debug("etask is nullptr");
```
isn't this case an infinite loop? I am not sure to understand

ikryukov replied:
it is error case, used it for debug, it should not happen in real situations

samnordmann replied:
Ok, why not using `ucc_assert` here then?

samnordmann commented on src/components/tl/cuda/bcast/bcast_linear.c:
```diff
@@ -0,0 +1,303 @@
+                // finish
+                task->bcast_linear.stage = STAGE_DONE;
+            }
+        case STAGE_DONE:
```
could we remove this stage?

ikryukov replied:
yes, we could, I left it to be more explicit

samnordmann reviewed (commented):
Looks good to me! Thanks!
I only left some minor remarks. Can you, in addition, add this algo to the tests?

ikryukov commented on src/components/tl/cuda/tl_cuda.h:
```diff
@@ -224,6 +225,17 @@ struct ucc_tl_cuda_task {
             size_t (*get_offset)(const ucc_tl_cuda_task_t *task,
                                  ucc_rank_t                block);
         } allgatherv_linear;
+        
```
done

ikryukov commented:
> Looks good to me! Thanks! I only left some minor remarks. Can you, in addition, add this algo to the tests?

Thanks for review, addressed comments and added test to validate bcast for cuda too.

ikryukov commented on src/components/tl/cuda/bcast/bcast_linear.c:
```diff
@@ -0,0 +1,296 @@
+
+    st = ucc_coll_task_get_executor(&task->super, &exec);
+    if (ucc_unlikely(st != UCC_OK)) {
+        return;
```
added

ikryukov commented on src/components/tl/cuda/tl_cuda.h:
```diff
@@ -224,6 +225,16 @@ struct ucc_tl_cuda_task {
+            ucc_rank_t              root;
+            size_t size;
+            int num_steps;
+            ucc_ee_executor_task_t *exec_task;
```
fixed

janjust reviewed (approved):
looks good

samnordmann commented:
ok to test

samnordmann commented on src/components/tl/cuda/bcast/bcast_linear.c:
```diff
@@ -0,0 +1,423 @@
+    task->bcast_linear.dt   = coll_args->args.src.info.datatype;
+    task->bcast_linear.sbuf = coll_args->args.src.info.buffer;
+
+    task->super.flags |= UCC_COLL_TASK_FLAG_EXECUTOR;
```
align the "="

samnordmann commented on src/components/tl/cuda/bcast/bcast_linear.c:
```diff
@@ -0,0 +1,423 @@
+        return status;
+    }
+
+    task->bcast_linear.root = coll_args->args.root;
+    task->bcast_linear.dt   = coll_args->args.src.info.datatype;
+    task->bcast_linear.sbuf = coll_args->args.src.info.buffer;
```
why is this needed?

samnordmann commented on src/components/tl/cuda/bcast/bcast_linear.c:
```diff
@@ -0,0 +1,423 @@
+        task->bcast_linear.stage = UCC_TL_TEAM_RANK(team) == task->bcast_linear.root ? STAGE_INIT_BAR_ROOT : STAGE_FIND_BAR_PEER;
+    }
+    
+    task->bcast_linear.size = ucc_dt_size(dt) * args->src.info.count;
+    task->bcast_linear.num_steps =
+        ucc_div_round_up(task->bcast_linear.size, half_scratch_size);
```
can this be moved to init?

samnordmann commented on src/components/tl/cuda/bcast/bcast_linear.c:
```diff
@@ -0,0 +1,423 @@
+              ucc_datatype_str(dt), task->bcast_linear.size,
+              task->bcast_linear.num_steps);
+
+    task->bcast_linear.sbuf = args->src.info.buffer;
```
duplicate with init

samnordmann commented on src/components/tl/cuda/bcast/bcast_linear.c:
```diff
@@ -0,0 +1,423 @@
+        if (ucc_tl_cuda_get_sync_root(task, task->bcast_linear.root) != UCC_OK) {
+            return;
+        }
+        task->bcast_linear.step = 0;
```
duplicate with start function

samnordmann commented on src/components/tl/cuda/bcast/bcast_linear.c:
```diff
@@ -0,0 +1,423 @@
+    }
+
+    switch (task->bcast_linear.stage) {
+    case STAGE_INIT_BAR_ROOT:
```
suggesting, not requesting: `STAGE_INIT_BAR_ROOT` and `STAGE_FIND_BAR_PEER` could be merged into one stage, with a if/else inside. This has the benefit of 1) reducing the number of cases, and, 2) allowing to fall through the next case `STAGE_SYNC` and thus saving some latency

ikryukov replied:
sure, need to test it, but I'd like to put it in separate PR

samnordmann commented on src/components/tl/cuda/bcast/bcast_linear.c:
```diff
@@ -0,0 +1,423 @@
+            etask = task->bcast_linear.exec_task;
+            ucc_assert(NULL != etask);
+            st = ucc_ee_executor_task_test(etask);
+            if (st == UCC_OK) {
+                ucc_ee_executor_task_finalize(etask);
+                task->bcast_linear.exec_task = NULL;
+                // signal others
+                ++task->bcast_linear.step;
+                set_rank_step(task, task->bcast_linear.root,
+                              task->bcast_linear.step, 0);
+                task->bcast_linear.stage = STAGE_WAIT_ALL;
+            } else {
+                // not ready
+                return;
+            }
```
Better readability to treat the `st != UCC_OK` case before (using that we can return), and also we need to handle if the executor is in error state
```suggestion
            if (st != UCC_OK) {
                task->super.status = st;
                return;
            }
            ucc_ee_executor_task_finalize(etask);
            task->bcast_linear.exec_task = NULL;
            // signal others
            ++task->bcast_linear.step;
            set_rank_step(task, task->bcast_linear.root,
                          task->bcast_linear.step, 0);
            task->bcast_linear.stage = STAGE_WAIT_ALL;
```

samnordmann commented on src/components/tl/cuda/bcast/bcast_linear.c:
```diff
@@ -0,0 +1,423 @@
+                // go to next iteration
+                task->bcast_linear.stage = STAGE_COPY;
+                return;
+            } else {
+                // finish
+                st = ucc_tl_cuda_shm_barrier_start(trank, task->bar);
+                if (ucc_unlikely(st != UCC_OK)) {
+                    ucc_error("failed to start barrier from root rank");
+                    task->super.status = st;
+                    return;
+                }
+                task->bcast_linear.stage = STAGE_WAIT_COMPLETION;
+            }
```
```suggestion
            }
            // finish
            st = ucc_tl_cuda_shm_barrier_start(trank, task->bar);
            if (ucc_unlikely(st != UCC_OK)) {
                ucc_error("failed to start barrier from root rank");
                task->super.status = st;
                return;
            }
            task->bcast_linear.stage = STAGE_WAIT_COMPLETION;
```

samnordmann commented on src/components/tl/cuda/bcast/bcast_linear.c:
```diff
@@ -0,0 +1,423 @@
+            ucc_tl_cuda_put_sync_root(task, task->bcast_linear.root);
+            task->super.status = UCC_OK;
+        default:
+            break;
```
```suggestion
            ucc_assert(0);
            break;
```

samnordmann commented on src/components/tl/cuda/bcast/bcast_linear.c:
```diff
@@ -0,0 +1,423 @@
+            etask = task->bcast_linear.exec_task;
+            ucc_assert(NULL != etask);
+            st = ucc_ee_executor_task_test(etask);
+            if (st == UCC_OK) {
```
better to save status and return if `st != UCC_OK` and take the rest of the code out of the if/then/else body.

samnordmann commented on src/components/tl/cuda/bcast/bcast_linear.c:
```diff
@@ -0,0 +1,423 @@
+                if (task->bcast_linear.step < task->bcast_linear.num_steps) {
+                    task->bcast_linear.stage = STAGE_WAIT_ROOT;
+                    return;
+                } else {
```
`else` not needed

samnordmann reviewed (commented):
Thanks Ilya! Only left minor comments

Note on copyright: at the very top of each file, there is a copyright header that needs to be updated with the current year (so I suggest 2025, anticipating a little bit). Each PR needs to comply with that...

samnordmann commented on src/components/tl/cuda/tl_cuda_coll.h:
```diff
@@ -100,19 +113,39 @@ ucc_status_t ucc_tl_cuda_task_init(ucc_base_coll_args_t *coll_args,
+        // root
+        if (task->subset.myrank == coll_args->args.root) {
+            int peer               = ucc_ep_map_eval(task->subset.map, 1);
+            task->bcast_linear.key = ((uint64_t)coll_args->args.tag << 32 |
```
also, I see that tag is only 16 bits:
https://github.com/openucx/ucc/blob/0c0fc21559835044ab107199e334f7157d6a0d3d/src/ucc/api/ucc_def.h#L135

So we could save space by shifting 48 bits instead of 32

samnordmann replied:
maybe it would be safer to also add some assert or static assert, that the sizes of the datatype do satisfy those underlying assumptions

ikryukov commented on src/components/tl/cuda/tl_cuda.h:
```diff
@@ -104,9 +106,12 @@ UCC_CLASS_DECLARE(ucc_tl_cuda_context_t, const ucc_base_context_params_t *,
 
 typedef uint32_t ucc_tl_cuda_sync_state_t;
 
+#define UCC_TAG_FREE 0xFFFFFFFFFFFFFFFF
```
currently it is used only by bcast algo only, so I'd prefer to narrow visibility

ikryukov commented on src/components/tl/cuda/tl_cuda_team.c:
```diff
@@ -268,14 +273,15 @@ ucc_status_t ucc_tl_cuda_team_create_test(ucc_base_team_t *tl_team)
     CUDA_CHECK_GOTO(cudaStreamCreateWithFlags(&team->stream,
                     cudaStreamNonBlocking), exit_err, status);
-    for (i = 0; i < lib->cfg.max_concurrent; i++) {
+    // second max_concurent events are unused for bcast active set
```
events in range [max_concurent; 2 * max_concurent) are reserved for active set

ikryukov commented on src/components/tl/cuda/bcast/bcast_linear.c:
```diff
@@ -0,0 +1,422 @@
+                                            ? (ucc_rank_t)task->subset.map.ep_num
+                                            : UCC_TL_TEAM_SIZE(team);
+    size_t              chunk_size        =
+                            task->bcast_linear.step < task->bcast_linear.num_steps
```
or `size_t chunk_size = ucc_min(scratch_size, task->bcast_linear.size - task->bcast_linear.step * scratch_size);`

janjust commented:
Need to see why checks are failing, but otherwise LGTM

janjust commented:
bot:retest

ikryukov commented on src/components/tl/cuda/bcast/bcast_linear.c:
```diff
@@ -0,0 +1,423 @@
+    int                        i;
+    ucc_status_t               st;
+
+    // Iterate over available barriers in active set pool to find a free one
```
we do it on init, in normal mode this state is unreachable -> we wouldn't call this functions

ikryukov commented on src/components/tl/cuda/bcast/bcast_linear.c:
```diff
@@ -0,0 +1,423 @@
+            st        = ucc_tl_cuda_shm_barrier_init_root(
+                task->subset.map.ep_num, task->subset.myrank,
+                task->bcast_linear.root, task->bar);
+            if (ucc_unlikely(st != UCC_OK)) {
```
right now yes, but current interface for barriers returns status, so need to check, we should address with during barrier refactoring (out of scope for this PR)
